autoload/ale/statusline.vim

```diff
@@ -1,29 +1,54 @@
 " Author: KabbAmine <amine.kabb@gmail.com>
 " Description: Statusline related function(s)
 
-function! ale#statusline#Status() abort
-    " Returns a formatted string that can be integrated in the
-    " statusline
-
-    let l:buffer = bufnr('%')
-    let l:loclist = g:ale_buffer_loclist_map
-
-    if !has_key(l:loclist, l:buffer)
-        return ''
-    endif
-
+" Update the buffer error/warning count with data from loclist.
+function! ale#statusline#Update(buffer, loclist) abort
     let l:errors = 0
     let l:warnings = 0
-    for l:entry in l:loclist[l:buffer]
+
+    for l:entry in a:loclist
         if l:entry.type ==# 'E'
             let l:errors += 1
         else
             let l:warnings += 1
         endif
     endfor
 
-    let l:errors = l:errors ? printf(g:ale_statusline_format[0], l:errors) : ''
-    let l:warnings = l:warnings ? printf(g:ale_statusline_format[1], l:warnings) : ''
+    let g:ale_buffer_count_map[a:buffer] = [l:errors, l:warnings]
+endfunction
+
+" Returns a tuple of errors and warnings (or false if no data exists)
+" for use in third-party integrations.
+function! ale#statusline#Count(buffer) abort
+    if !has_key(g:ale_buffer_count_map, a:buffer)
+        if has_key(g:ale_buffer_loclist_map, a:buffer)
+            call ale#statusline#Update(a:buffer, g:ale_buffer_loclist_map[a:buffer])
+            return ale#statusline#Count(a:buffer)
+        else
+            return 0
+        endif
+    endif
+
+    return g:ale_buffer_count_map[a:buffer]
+endfunction
+
+" Returns a formatted string that can be integrated in the statusline.
+function! ale#statusline#Status() abort
+    let l:buffer = bufnr('%')
+
+    if !has_key(g:ale_buffer_count_map, l:buffer)
+        if has_key(g:ale_buffer_loclist_map, l:buffer)
+            call ale#statusline#Update(l:buffer, g:ale_buffer_loclist_map[l:buffer])
+            return ale#statusline#Status()
+        else
+            return ''
+        endif
+    endif
```

Comment:
Then the return value here will always be consistent between when ALE is first loaded, and when it reports no errors.